File: phase-1-dsa-foundations/module-2-core-data-structures/trees/TreeOperations.java
package trees;

import java.util.*;

/**
 * Demonstrating Binary Tree Logic and Traversals.
 */
public class TreeOperations {

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    public static void main(String[] args) {
        System.out.println("--- Tree Operations ---");

        // 1. Creation
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        /*
         * 1
         * / \
         * 2 3
         * / \
         * 4 5
         */

        // 2. DFS Traversals
        System.out.print("Pre-Order: ");
        preOrder(root);
        System.out.println();

        System.out.print("In-Order:  ");
        inOrder(root);
        System.out.println();

        System.out.print("Post-Order:");
        postOrder(root);
        System.out.println();

        // 3. BFS Traversal
        System.out.print("BFS/Level: ");
        levelOrder(root);
        System.out.println();
    }

    // Root -> Left -> Right
    private static void preOrder(TreeNode node) {
        if (node == null)
            return;
        System.out.print(node.val + " ");
        preOrder(node.left);
        preOrder(node.right);
    }

    // Left -> Root -> Right
    private static void inOrder(TreeNode node) {
        if (node == null)
            return;
        inOrder(node.left);
        System.out.print(node.val + " ");
        inOrder(node.right);
    }

    // Left -> Right -> Root
    private static void postOrder(TreeNode node) {
        if (node == null)
            return;
        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.val + " ");
    }

    // Level by Level using Queue
    private static void levelOrder(TreeNode root) {
        if (root == null)
            return;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode curr = queue.poll();
            System.out.print(curr.val + " ");
            if (curr.left != null)
                queue.offer(curr.left);
            if (curr.right != null)
                queue.offer(curr.right);
        }
    }
}

File: go/phase-1-dsa-foundations/module-2-core-data-structures/trees/tree_operations.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// DFS: Pre-Order (Root -> Left -> Right)
func PreOrder(node *TreeNode) {
	if node == nil {
		return
	}
	fmt.Printf("%d ", node.Val)
	PreOrder(node.Left)
	PreOrder(node.Right)
}

// DFS: In-Order (Left -> Root -> Right)
func InOrder(node *TreeNode) {
	if node == nil {
		return
	}
	InOrder(node.Left)
	fmt.Printf("%d ", node.Val)
	InOrder(node.Right)
}

// DFS: Post-Order (Left -> Right -> Root)
func PostOrder(node *TreeNode) {
	if node == nil {
		return
	}
	PostOrder(node.Left)
	PostOrder(node.Right)
	fmt.Printf("%d ", node.Val)
}

// BFS: Level Order
func LevelOrder(root *TreeNode) {
	if root == nil {
		return
	}
	
	queue := []*TreeNode{root}
	
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:] // Dequeue
		
		fmt.Printf("%d ", curr.Val)
		
		if curr.Left != nil {
			queue = append(queue, curr.Left)
		}
		if curr.Right != nil {
			queue = append(queue, curr.Right)
		}
	}
}

func main() {
	fmt.Println("--- Tree Operations ---")
	
	// Create Tree
	//      1
	//     / \
	//    2   3
	//   / \
	//  4   5
	root := &TreeNode{Val: 1}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 3}
	root.Left.Left = &TreeNode{Val: 4}
	root.Left.Right = &TreeNode{Val: 5}
	
	fmt.Print("Pre-Order:  ")
	PreOrder(root)
	fmt.Println()
	
	fmt.Print("In-Order:   ")
	InOrder(root)
	fmt.Println()
	
	fmt.Print("Post-Order: ")
	PostOrder(root)
	fmt.Println()
	
	fmt.Print("Level-Order:")
	LevelOrder(root)
	fmt.Println()
}
